src/main/java/com/google/devtools/build/lib/actions/Artifact.java

```diff
@@ -563,12 +566,17 @@ public SpecialArtifact getParent() {
    *
    * <p> The directory name is always a relative path to the execution directory.
    */
-  @Override
   public final String getDirname() {
     PathFragment parent = execPath.getParentDirectory();
     return (parent == null) ? "/" : parent.getSafePathString();
   }
 
+  @Override
+  public final String getDirnameForStarlark(StarlarkSemantics starlarkSemantics)
+      throws EvalException {
+    return starlarkSemantics.get(ArtifactPathMapper.STARLARK_SEMANTICS_KEY).getMappedDirname(this);
```

Comment:
Not sure I'm reading this right: is this providing the ability for `dirname` to resolve to mapped paths? How does that fit with

> must not contain `dirname`, ...

?